pkg/controller/job/job_controller.go

```diff
@@ -1742,6 +1747,11 @@ func isPodFailed(p *v1.Pod, job *batch.Job, wFinalizers bool) bool {
 		// TODO(#113855): Stop limiting this behavior to Jobs with podFailurePolicy.
 		// For now, we do so to avoid affecting all running Jobs without the
 		// avaibility to opt-out into the old behavior.
+		if p.Status.Phase == v1.PodPending {
+			_, _, action := matchPodFailurePolicy(job.Spec.PodFailurePolicy, p)
+			failJob := batch.PodFailurePolicyActionFailJob
+			return action != nil && *action == failJob
```

Comment:
doesn't this mean that there is no timeout at all?
That as soon as the pod has the condition, even if transient, it matches the policy?
I don't think we want that.

Reply:
What should be the next steps?  I think I am hearing that we should include a deadline for PodFailurePolicy to match on a particular condition.  

Is it possible to add this to the retriable jobs KEP as an enhancement found during Beta?  Or is this a separate KEP on top of a beta feature?

Reply:
I think there should be a mechanism that, after a certain time has passed, marks a Pending pod as Failed.

As discussed in the WG Batch meeting, this **could be** an external controller. But it could also be kubelet if there is an API for it.

That's definitely material for a separate KEP. But I would suggest bringing up this discussion back to SIG Node again.

Reply:
Transitioning Pending pods with `DeletionTimestamp!= nil` is going to be done as part of the [KEP-3329: Retriable and non-retriable Pod failures for Jobs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures#ga): "Also, extend PodGC to mark as failed (and delete) terminating pods (with set deletionTimestamp) which stuck in the pending.". We planned to do this in PodGC. However, this only refers to Pending pods which are deleted, so not sure it applies in your case.

Reply:
Right, so the external controller can simply issue a DELETE based on its timeout.